### PR TITLE
add music packs support

### PR DESCRIPTION
**Pull request by rfomin**
Currently it just loads music files from the `autoload`. We can also add a `music_packs` directory like Choco/Crispy. How about adding support for .zip files?

### CODE REVIEW

**Comment by fabiangreffrath** [comment]
> Currently it just loads music files from the `autoload`. 

So, it simply adds the music files as if they were LMP files?

> We can also add a `music_packs` directory like Choco/Crispy. 

Not sure if this is necessary, I always found this a bit cumbersome.

> How about adding support for .zip files?

Sure, we have miniz so we can do that! The question is if we support lumps inside a ZIP or even WADs inside a ZIP? Do we keep copies of the uncompressed file sin memory or write them to disc temporarily?

**Comment by rfomin** [comment]
> So, it simply adds the music files as if they were LMP files?

Yes.

> The question is if we support lumps inside a ZIP or even WADs inside a ZIP? Do we keep copies of the uncompressed files in memory or write them to disc temporarily?

I think we can support WAD, lumps and music files. Temporary writing to disk is easier since we can just use our existing `AutoLoadWADs` function, but decompressing files in memory is probably cleaner, not sure.

**Comment by fabiangreffrath** [comment]
> but decompressing files in memory is probably cleaner, not sure.

I haven't dones either yet, just single lumps in memory (ZDBSP nodes).

**Comment by fabiangreffrath** [comment]
Doesn't seem to be *that* tricky:
https://github.com/richgel999/miniz/blob/76b3a872855388c735c564905da030f26334f3b3/examples/example2.c#L117

**Comment by rfomin** [comment]
> Doesn't seem to be _that_ tricky:

I tried temporary directory. It's simple, but we can't remove the temporary directory when exiting the program because we never close file handlers in w_wad.c Perhaps an in-memory solution is better.

**Comment by fabiangreffrath** [comment]
> > Doesn't seem to be _that_ tricky:
> 
> I tried temporary directory. It's simple, but we can't remove the temporary directory when exiting the program because we never close file handlers in `w_wad.c` Perhaps an in-memory solution is better.

We currently don't have any means to keep data entirely in memory, like e.g. memmap(). But we could register an atexit() handle that closes all file handlers in all temporary directories and removes these directories afterwards.

**Comment by fabiangreffrath** [comment]
I guess we would just need to remember all the temp directories that we created and then glob through them and delete all the files. We could factor out the code into w_zip.c if you like.

**Comment by rfomin** [comment]
> I guess we would just need to remember all the temp directories that we created and then glob through them and delete all the files.

Yes, that's what I did. Ready for review.

File: src/d_main.c [hunk 1]
```diff
@@ -571,6 +573,74 @@ void D_StartTitle (void)
...
 static char title[128];
 
+char **tempdirs = NULL;
+
+static void AutoLoadWADs(const char *path);
+
+static boolean D_AddZipFile(const char *file)
+{
+  int i;
+  mz_zip_archive  zip_archive;
```
Comment (fabiangreffrath) [inline src/d_main.c:583]: double space

File: src/d_main.c [hunk 2]
```diff
@@ -571,6 +573,74 @@ void D_StartTitle (void)
...
+    }
+
+    free(dest);
+  }
+
+  mz_zip_reader_end(&zip_archive);
+
+  AutoLoadWADs(tempdir);
+
+  tempdirs = I_Realloc(tempdirs, (idx + 2) * sizeof(*tempdirs));
```
Comment (fabiangreffrath) [inline src/d_main.c:637]: why `+2` not `+1`?

File: src/d_main.c [hunk 3]
```diff
@@ -571,6 +573,74 @@ void D_StartTitle (void)
...
+  if (!mz_zip_reader_init_file(&zip_archive, file, MZ_ZIP_FLAG_DO_NOT_SORT_CENTRAL_DIRECTORY))
+  {
+    printf("D_AddZipFile: Failed to open %s\n", file);
+    free(tempdir);
+    return true;
+  }
+
+  M_MakeDirectory(tempdir);
+
+  for (i = 0; i < (int)mz_zip_reader_get_num_files(&zip_archive); ++i)
```
Comment (fabiangreffrath) [inline src/d_main.c:612]: Can we evaluate this only once please?

File: src/d_main.c [hunk 4]
```diff
@@ -571,6 +573,74 @@ void D_StartTitle (void)
...
+
+  if (!M_StringEndsWith(str, ".zip"))
+  {
+    free(str);
+    return false;
+  }
+
+  free(str);
+
+  str = M_StringJoin("_woof_", M_BaseName(file), NULL);
```
[the remark this answers is not in the document]
Comment (rfomin) [reply src/d_main.c:598]: Just use the zip filename here? This is the name for the temporary directory.

File: src/d_main.c [hunk 5]
```diff
@@ -1394,7 +1467,7 @@ static void AutoLoadWADs(const char *path)
     const char *filename;
 
     glob = I_StartMultiGlob(path, GLOB_FLAG_NOCASE|GLOB_FLAG_SORTED,
-                            "*.wad", "*.lmp", NULL);
```
Comment (fabiangreffrath) [inline src/d_main.c:1397]: `*.zip` as well now?

Comment (rfomin) [reply src/d_main.c:1397]: No, we extract `*.zip` in `D_AddFile`.

File: src/d_quit.c [hunk 6]
```diff
@@ -16,33 +16,80 @@
...
+
+    W_CloseFileDescriptors();
+
+    for (i = 0; tempdirs[i]; ++i)
+    {
+        glob_t *glob;
+        const char *filename;
+
+        glob = I_StartMultiGlob(tempdirs[i], GLOB_FLAG_NOCASE|GLOB_FLAG_SORTED,
+                                "*.*", NULL);
```
[the remark this answers is not in the document]
Comment (rfomin) [reply src/d_quit.c:71]: I think we should extract only .wad, .lmp and music files. I'll fix this in `D_AddZipFile`

File: src/d_main.c (hunk 3 above)
Comment (rfomin) [reply src/d_main.c:612]: It returns constant https://github.com/fabiangreffrath/woof/blob/9aa0dbe1d2b92047cc1322b3eeefb8fec3f4124e/miniz/miniz.c#L7656

**Comment by fabiangreffrath** [comment]
How about DEH/BEX files, so we could load doomworld/dsda archives right after downloading? 😮

File: src/d_main.c [hunk 7]
```diff
@@ -571,6 +573,81 @@ void D_StartTitle (void)
...
+    mz_zip_archive_file_stat file_stat;
+    char *name;
+
+    mz_zip_reader_file_stat(&zip_archive, i, &file_stat);
+
+    if (file_stat.m_is_directory)
+      continue;
+
+    name = M_StringDuplicate(M_BaseName(file_stat.m_filename));
+    M_ForceLowercase(name);
```
Comment (fabiangreffrath) [inline src/d_main.c:619]: Should we probably introduce a case-insensitive variant of `M_StringEndsWith()`?

**Comment by rfomin** [comment]
> How about DEH/BEX files, so we could load doomworld/dsda archives right after downloading? 😮

90% of PWADs have a built-in DEH file. Maybe some older pre-2000 PWADs need it, but they usually have a complicated setup procedure anyway. I checked all my downloads and they are already working.

File: src/d_main.c (hunk 7 above)
Comment (rfomin) [reply src/d_main.c]: Done in https://github.com/fabiangreffrath/woof/pull/649/commits/4a8b277a56c7c9e0fb0d9a0ee57ba26958ba1ef5

**Comment by fabiangreffrath** [comment]
> 90% of PWADs have a built-in DEH file. Maybe some older pre-2000 PWADs need it, but they usually have a complicated setup procedure anyway. I checked all my downloads and they are already working.

Hm, I thought this would be the most obvious expansion of the concept. 🤔

**Comment by fabiangreffrath** [comment]
> But what about the load order? Also, many vanilla PWADs have a built-in DEHACKED lump and a DEH file for Choco, usually loading both doesn't harm, it just seems excessive. I believe that there will be a lot of code that will still not work in all cases.

Yes, okay.

File: src/d_main.c [hunk 8]
```diff
@@ -1203,7 +1276,8 @@ static int GuessFileType(const char *name)
         iwad_found = true;
     }
     else if (M_StringEndsWith(lower, ".wad") ||
-             M_StringEndsWith(lower, ".lmp"))
+             M_StringEndsWith(lower, ".lmp") ||
+             M_StringEndsWith(lower, ".zip"))
```
Comment (fabiangreffrath) [inline src/d_main.c:1280]: Could you also add zip to the bash completion file?

File: src/CMakeLists.txt [hunk 9]
```diff
@@ -183,9 +183,7 @@ if("${${PROJECT_NAME}_STRICT}")
     target_compile_definitions(woof PRIVATE MBF_STRICT)
 endif()
 
-target_compile_definitions(woof PRIVATE MINIZ_NO_STDIO)
 target_compile_definitions(woof PRIVATE MINIZ_NO_TIME)
-target_compile_definitions(woof PRIVATE MINIZ_NO_ARCHIVE_APIS)
```
[the remark this answers is not in the document]
Comment (rfomin) [reply src/CMakeLists.txt:188]: It seems we can't because there are warnings in linux clang and msys2 gcc.

File: src/d_main.c (hunk 8 above)
Comment (rfomin) [reply src/d_main.c:1280]: Done in https://github.com/fabiangreffrath/woof/pull/649/commits/eae2bc0e9b31ea7d30e8b197ce038e6dbe8e96b1

**Comment by fabiangreffrath** [comment]
Very cool feature, thank you!

**Pull request merged**